### PR TITLE
Patch up fetching dates from source

### LINKED ISSUE #69382: failed to parse date field after upgrade to 7.11.x from 7.10.2
We are getting the following error after upgrading from 7.10.2 to 7.11.0 (subsequent update to 7.11.1 did not resolve the issue).

`failed to parse date field [1.613970929E12] with format [strict_date_optional_time||epoch_millis]`

Mapping is basic and remains unchanged from 7.10.2 configuration. 

Fields throwing the error.... 
```       
        "ProcTime": {
          "type": "date",
          "store": true
        },
        "RcvTime": {
          "type": "date",
          "store": true
        }
```

We have multiple deployments and only the ones upgraded to 7.11.x are having this issue. 

Since nothing on our end changed and I didn't see anything in the 7.11.x breaking changes or release notes I am posting here.  Please let me know what I can provide to assist with reproducing or identifying the issue. 

Thank you

### PR DESCRIPTION
**Pull request by nik9000**
This fixes an issue that `fields` has with dates sent in the
`###.######` format.

If you send a date in the format `#####.######` we'll parse the bit
before the decimal place as the number of milliseconds since epoch and
we'll parse the bit after the decimal as the number of nanoseconds since
the start of that millisecond. This works and is convenient for some
folks. Sadly, the code that back the `fields` API for dates doesn't work
with the string format in this case - it works with a `double`. `double`
is bad for two reasons:
1. It's default string representation is scientific notation and our
   parsers don't know how to deal with that.
2. It loses precision relative to the string representation. `double`
   only has 52 bits of mantissa which can precisely store the number of
   nanoseconds until about 6am on April 15th, 1970. After that it starts
   to lose precision.

This fixed the first issue, getting us the correct string
representation is a "quick and dirty" way. It just converts the `double`
back to a string. But we still lose precision. Fixing that would require
a larger change.....

Closes #69382

### CODE REVIEW

**Comment by nik9000** [comment]
I am 10000000% sure this is the wrong way to fix this. But it stops us from ignoring dates sent as fixed point numbers. We just sometimes get rounding errors.....

**Review by pgomulka — COMMENTED** [review]
I wonder if the loose of precision is acceptable (although I also can not come up with a better solution)..
we actually allow 19digis for epoch seconds and 9digits for nanoseconds..

a wild idea.. Are we able to configure jackson to parse floats as bigdecimals or strings? there is something like DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS
Or even something more complex like custom json deserialiser and force parsing doubles as strings only for date fields (totally no idea if that is possible)

**Comment by nik9000** [comment]
I imagine the loss of precision isn't acceptable. But maybe it's acceptable
in the short term. I thing changing SourceLookup to return big decimal
would be kind of breaky. Scripts get those numbers. Update gets them. I
imagine others do too.

I was thinking it might be better to try and move off of the "to map"
parsing so we can share more code with the from xconent parsing. But that's
hard! The code doesn't really want to be shared and the callers don't
expect it.

**Comment by nik9000** [comment]
I've split up the tests for easier failure diagnosis and to make it easier to add `@Repeat` to individual tests. I believe I've worked through the randomized failures as they stand now. In the future I'd like to force all subclasses of `MapperTestCase` to implement this sort of method but I worry that it'd make the change too large to backport as far as we might like. And I worry that it'll fail randomly in old branches and be a nuisance. So I'd like to do that work in a follow up.

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Review by pgomulka — APPROVED** [review]
LGTM, thanks for the fix. I don't think we can come up with anything better loosing precision short term

File: server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java [hunk 1]
```diff
@@ -355,7 +373,8 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
             return new SourceValueFetcher(name(), context, nullValue) {
                 @Override
                 public String parseSourceValue(Object value) {
-                    String date = value.toString();
+                    String date = value instanceof Number ? NUMBER_FORMAT.format(value) : value.toString();
+                    // TODO can we emit a warning if we're losing precision here? I'm not sure we can.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:377]: Yeah! I was more thinking about how to detect that we're losing precision. I think it's probably safest to leave it for a follow up change.

**Pull request merged**